src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs

```diff
@@ -427,6 +460,7 @@ static void Main()
             Assert.Null(model.GetTypeInfo(def).Type);
             Assert.Null(model.GetSymbolInfo(def).Symbol);
             Assert.Null(model.GetDeclaredSymbol(def));
+            Assert.Equal(0, model.GetConstantValue(def).Value);
```

Comment:
The ```null``` literal is special, it will only "match" reference types and will match any reference type. So, its constant value is null regardless of the target type. As in these scenarios, for example:
```
        [Fact]
        public void AssignNullToInt()
        {
            string source = @"
class C
{
    static void Main()
    {
        int x = null;
        System.Console.Write(x);
    }
}
";
            var comp = CreateCompilationWithMscorlib(source, parseOptions: TestOptions.ExperimentalParseOptions);

            var tree = comp.SyntaxTrees.First();
            var model = comp.GetSemanticModel(tree);
            var def = tree.GetCompilationUnitRoot().DescendantNodes().OfType<LiteralExpressionSyntax>().Single();

            Assert.Equal("null", def.ToString());
            Assert.True(model.GetConstantValue(def).HasValue);
            Assert.Null(model.GetConstantValue(def).Value);
            Assert.Equal("System.Int32", model.GetTypeInfo(def).ConvertedType.ToTestDisplayString());
            Assert.Null(model.GetTypeInfo(def).Type);
        }

        [Fact]
        public void AssignNullToNullableInt()
        {
            string source = @"
class C
{
    static void Main()
    {
        int? x = null;
        System.Console.Write(x);
    }
}
";
            var comp = CreateCompilationWithMscorlib(source, parseOptions: TestOptions.ExperimentalParseOptions);

            var tree = comp.SyntaxTrees.First();
            var model = comp.GetSemanticModel(tree);
            var def = tree.GetCompilationUnitRoot().DescendantNodes().OfType<LiteralExpressionSyntax>().Single();

            Assert.Equal("null", def.ToString());
            Assert.True(model.GetConstantValue(def).HasValue);
            Assert.Null(model.GetConstantValue(def).Value);
            Assert.Equal("System.Int32?", model.GetTypeInfo(def).ConvertedType.ToTestDisplayString());
            Assert.Null(model.GetTypeInfo(def).Type);
        }
```

Clearly null value doesn't match the ConvertedType in cases like this, its the value of the null literal regardless of the target type.

However, if we look at literals of value types:
```
        [Fact]
        public void AssignIntToLong()
        {
            string source = @"
class C
{
    static void Main()
    {
        long x = 1;
        System.Console.Write(x);
    }
}
";
            var comp = CreateCompilationWithMscorlib(source, parseOptions: TestOptions.ExperimentalParseOptions);
            comp.VerifyDiagnostics();

            var tree = comp.SyntaxTrees.First();
            var model = comp.GetSemanticModel(tree);
            var def = tree.GetCompilationUnitRoot().DescendantNodes().OfType<LiteralExpressionSyntax>().Single();

            Assert.Equal(1, model.GetConstantValue(def).Value);
            Assert.Equal(typeof(int), model.GetConstantValue(def).Value.GetType());
            Assert.Equal("System.Int64", model.GetTypeInfo(def).ConvertedType.ToTestDisplayString());
            Assert.Equal("System.Int32", model.GetTypeInfo(def).Type.ToTestDisplayString());
        }
```
We can observe that the constant value matches the Type, not the ConvertedType. 

I understand that ```default``` is special in the way that it doesn't have natural type (similar to ```null```). At the same time it doesn't have a "natural" constant value that is going to be valid for any target type in an non-error scenario. In fact, for some target types there could be no valid constant value. Given that, it would feel right to me If the Type (not Converted type) and the ConstantValue of the node were both based on the target type. But I could be convinced that the Type should always bee null. I would still like to avoid any special cases in SemanticModel to get this behavior.